File: c/algorithms/numeric.c
/*
 * Numeric Algorithms Implementation in C
 * Includes: Primality test, base conversions, GCD, factorial, etc.
 */

#include <stdio.h>
#include <stdlib.h>
#include <stdbool.h>
#include <math.h>
#include <string.h>

// 9.1 Primality Test
bool is_prime(int n) {
    if (n < 2) return false;
    if (n == 2) return true;
    if (n % 2 == 0) return false;
    
    for (int i = 3; i <= sqrt(n); i += 2) {
        if (n % i == 0) return false;
    }
    
    return true;
}

void sieve_of_eratosthenes(int limit, int* primes, int* count) {
    if (limit < 2) {
        *count = 0;
        return;
    }
    
    bool* is_prime_arr = (bool*)malloc((limit + 1) * sizeof(bool));
    for (int i = 0; i <= limit; i++) {
        is_prime_arr[i] = true;
    }
    is_prime_arr[0] = is_prime_arr[1] = false;
    
    for (int i = 2; i <= sqrt(limit); i++) {
        if (is_prime_arr[i]) {
            for (int j = i * i; j <= limit; j += i) {
                is_prime_arr[j] = false;
            }
        }
    }
    
    *count = 0;
    for (int i = 2; i <= limit; i++) {
        if (is_prime_arr[i]) {
            primes[(*count)++] = i;
        }
    }
    
    free(is_prime_arr);
}

// 9.2 Base Conversions
void decimal_to_base(int number, int base, char* result) {
    if (number == 0) {
        strcpy(result, "0");
        return;
    }
    
    char digits[] = "0123456789ABCDEFGHIJKLMNOPQRSTUVWXYZ";
    char temp[100];
    int index = 0;
    bool negative = number < 0;
    number = abs(number);
    
    while (number > 0) {
        temp[index++] = digits[number % base];
        number /= base;
    }
    
    int result_index = 0;
    if (negative) result[result_index++] = '-';
    
    for (int i = index - 1; i >= 0; i--) {
        result[result_index++] = temp[i];
    }
    result[result_index] = '\0';
}

int base_to_decimal(const char* number_str, int base) {
    char digits[] = "0123456789ABCDEFGHIJKLMNOPQRSTUVWXYZ";
    int result = 0;
    int power = 0;
    int len = strlen(number_str);
    bool negative = (number_str[0] == '-');
    int start = negative ? 1 : 0;
    
    for (int i = len - 1; i >= start; i--) {
        char c = number_str[i];
        if (c >= 'a' && c <= 'z') c -= 32; // Convert to uppercase
        
        char* pos = strchr(digits, c);
        int value = pos - digits;
        result += value * pow(base, power++);
    }
    
    return negative ? -result : result;
}

// 9.3 Greatest Common Denominator (GCD)
int gcd(int a, int b) {
    a = abs(a);
    b = abs(b);
    
    while (b != 0) {
        int temp = b;
        b = a % b;
        a = temp;
    }
    
    return a;
}

int gcd_recursive(int a, int b) {
    a = abs(a);
    b = abs(b);
    
    if (b == 0) return a;
    return gcd_recursive(b, a % b);
}

int lcm(int a, int b) {
    return abs(a * b) / gcd(a, b);
}

// 9.4 Maximum value for base N with D digits
long long max_value_for_base(int base, int num_digits) {
    return (long long)pow(base, num_digits) - 1;
}

// 9.5 Factorial
long long factorial(int n) {
    if (n < 0) return -1; // Error
    
    long long result = 1;
    for (int i = 2; i <= n; i++) {
        result *= i;
    }
    
    return result;
}

long long factorial_recursive(int n) {
    if (n < 0) return -1;
    if (n == 0 || n == 1) return 1;
    
    return n * factorial_recursive(n - 1);
}

// Additional: Fibonacci
long long fibonacci(int n) {
    if (n < 0) return -1;
    if (n == 0) return 0;
    if (n == 1) return 1;
    
    long long a = 0, b = 1;
    for (int i = 2; i <= n; i++) {
        long long temp = a + b;
        a = b;
        b = temp;
    }
    
    return b;
}

// Additional: Fast Exponentiation
long long power(long long base, int exponent) {
    if (exponent == 0) return 1;
    
    bool negative = exponent < 0;
    exponent = abs(exponent);
    
    long long result = 1;
    long long current_power = base;
    
    while (exponent > 0) {
        if (exponent % 2 == 1) {
            result *= current_power;
        }
        current_power *= current_power;
        exponent /= 2;
    }
    
    return negative ? 1.0 / result : result;
}

// Main function for testing
int main() {
    printf("=== Numeric Algorithms Demo (C) ===\n\n");
    
    // Primality Test
    printf("1. Primality Test:\n");
    int test_numbers[] = {2, 3, 4, 17, 20, 29, 100, 97};
    for (int i = 0; i < 8; i++) {
        printf("  %d is %s\n", test_numbers[i], 
               is_prime(test_numbers[i]) ? "prime" : "not prime");
    }
    
    int primes[100];
    int count;
    sieve_of_eratosthenes(50, primes, &count);
    printf("\n  Primes up to 50: ");
    for (int i = 0; i < count; i++) {
        printf("%d ", primes[i]);
    }
    printf("\n");
    
    // Base Conversions
    printf("\n2. Base Conversions:\n");
    char result[100];
    decimal_to_base(255, 2, result);
    printf("  255 in binary: %s\n", result);
    decimal_to_base(255, 16, result);
    printf("  255 in hexadecimal: %s\n", result);
    decimal_to_base(255, 8, result);
    printf("  255 in octal: %s\n", result);
    
    printf("  Binary 11111111 to decimal: %d\n", base_to_decimal("11111111", 2));
    printf("  Hex FF to decimal: %d\n", base_to_decimal("FF", 16));
    
    // GCD and LCM
    printf("\n3. GCD and LCM:\n");
    printf("  GCD(48, 18) = %d\n", gcd(48, 18));
    printf("  GCD(100, 75) = %d\n", gcd(100, 75));
    printf("  LCM(12, 18) = %d\n", lcm(12, 18));
    
    // Maximum value for base
    printf("\n4. Maximum Value for Base:\n");
    printf("  Max value for base 10, 3 digits: %lld\n", max_value_for_base(10, 3));
    printf("  Max value for base 2, 8 digits: %lld\n", max_value_for_base(2, 8));
    printf("  Max value for base 16, 2 digits: %lld\n", max_value_for_base(16, 2));
    
    // Factorial
    printf("\n5. Factorial:\n");
    for (int i = 0; i <= 10; i++) {
        printf("  %d! = %lld\n", i, factorial(i));
    }
    
    // Fibonacci
    printf("\n6. Fibonacci Sequence:\n");
    printf("  First 15 Fibonacci numbers: ");
    for (int i = 0; i < 15; i++) {
        printf("%lld ", fibonacci(i));
    }
    printf("\n");
    
    // Fast Exponentiation
    printf("\n7. Fast Exponentiation:\n");
    printf("  2^10 = %lld\n", power(2, 10));
    printf("  5^3 = %lld\n", power(5, 3));
    
    return 0;
}

File: python/algorithms/numeric.py
"""
Numeric Algorithms Implementation
Includes: Primality test, base conversions, GCD, factorial, etc.
"""

import math

# 9.1 Primality Test
def is_prime(n):
    """
    Check if a number is prime - O(√n) time
    A prime number is only divisible by 1 and itself
    """
    if n < 2:
        return False
    if n == 2:
        return True
    if n % 2 == 0:
        return False
    
    # Check odd divisors up to √n
    for i in range(3, int(math.sqrt(n)) + 1, 2):
        if n % i == 0:
            return False
    
    return True


def sieve_of_eratosthenes(limit):
    """
    Find all primes up to limit - O(n log log n) time
    More efficient for finding multiple primes
    """
    if limit < 2:
        return []
    
    # Create boolean array, initially all True
    is_prime_arr = [True] * (limit + 1)
    is_prime_arr[0] = is_prime_arr[1] = False
    
    for i in range(2, int(math.sqrt(limit)) + 1):
        if is_prime_arr[i]:
            # Mark all multiples as not prime
            for j in range(i * i, limit + 1, i):
                is_prime_arr[j] = False
    
    # Collect all prime numbers
    return [i for i in range(limit + 1) if is_prime_arr[i]]


# 9.2 Base Conversions
def decimal_to_base(number, base):
    """
    Convert decimal number to any base (2-36) - O(log n) time
    """
    if number == 0:
        return "0"
    
    if base < 2 or base > 36:
        raise ValueError("Base must be between 2 and 36")
    
    digits = "0123456789ABCDEFGHIJKLMNOPQRSTUVWXYZ"
    result = []
    negative = number < 0
    number = abs(number)
    
    while number > 0:
        remainder = number % base
        result.append(digits[remainder])
        number //= base
    
    if negative:
        result.append('-')
    
    return ''.join(reversed(result))


def base_to_decimal(number_str, base):
    """
    Convert number from any base to decimal - O(n) time
    """
    if base < 2 or base > 36:
        raise ValueError("Base must be between 2 and 36")
    
    digits = "0123456789ABCDEFGHIJKLMNOPQRSTUVWXYZ"
    number_str = number_str.upper()
    
    result = 0
    power = 0
    
    for digit in reversed(number_str):
        if digit == '-':
            result = -result
        else:
            value = digits.index(digit)
            if value >= base:
                raise ValueError(f"Invalid digit '{digit}' for base {base}")
            result += value * (base ** power)
            power += 1
    
    return result


# 9.3 Greatest Common Denominator (GCD)
def gcd(a, b):
    """
    Euclidean algorithm for GCD - O(log min(a,b)) time
    """
    a, b = abs(a), abs(b)
    
    while b != 0:
        a, b = b, a % b
    
    return a


def gcd_recursive(a, b):
    """
    Recursive version of GCD
    """
    a, b = abs(a), abs(b)
    
    if b == 0:
        return a
    return gcd_recursive(b, a % b)


def lcm(a, b):
    """
    Least Common Multiple using GCD - O(log min(a,b)) time
    LCM(a,b) = (a * b) / GCD(a,b)
    """
    return abs(a * b) // gcd(a, b)


# 9.4 Maximum value for base N with D digits
def max_value_for_base(base, num_digits):
    """
    Compute maximum value for a number of specific base with N digits
    For base B with D digits: (B^D) - 1
    Example: Base 10, 3 digits = 999 = (10^3) - 1
    """
    if base < 2:
        raise ValueError("Base must be at least 2")
    if num_digits < 1:
        raise ValueError("Number of digits must be at least 1")
    
    return (base ** num_digits) - 1


# 9.5 Factorial
def factorial(n):
    """
    Iterative factorial - O(n) time, O(1) space
    """
    if n < 0:
        raise ValueError("Factorial not defined for negative numbers")
    
    result = 1
    for i in range(2, n + 1):
        result *= i
    
    return result


def factorial_recursive(n):
    """
    Recursive factorial - O(n) time, O(n) space
    """
    if n < 0:
        raise ValueError("Factorial not defined for negative numbers")
    
    if n == 0 or n == 1:
        return 1
    
    return n * factorial_recursive(n - 1)


def factorial_memoized(n, memo=None):
    """
    Memoized factorial for better performance with repeated calls
    """
    if memo is None:
        memo = {}
    
    if n < 0:
        raise ValueError("Factorial not defined for negative numbers")
    
    if n in memo:
        return memo[n]
    
    if n == 0 or n == 1:
        return 1
    
    memo[n] = n * factorial_memoized(n - 1, memo)
    return memo[n]


# Additional numeric algorithms
def fibonacci(n):
    """
    Compute nth Fibonacci number - O(n) time, O(1) space
    """
    if n < 0:
        raise ValueError("Fibonacci not defined for negative numbers")
    if n == 0:
        return 0
    if n == 1:
        return 1
    
    a, b = 0, 1
    for _ in range(2, n + 1):
        a, b = b, a + b
    
    return b


def power(base, exponent):
    """
    Fast exponentiation using binary method - O(log n) time
    """
    if exponent == 0:
        return 1
    if exponent < 0:
        return 1 / power(base, -exponent)
    
    result = 1
    current_power = base
    
    while exponent > 0:
        if exponent % 2 == 1:
            result *= current_power
        current_power *= current_power
        exponent //= 2
    
    return result


# Example usage and testing
if __name__ == "__main__":
    print("=== Numeric Algorithms Demo ===\n")
    
    # Primality Test
    print("1. Primality Test:")
    test_numbers = [2, 3, 4, 17, 20, 29, 100, 97]
    for num in test_numbers:
        print(f"  {num} is {'prime' if is_prime(num) else 'not prime'}")
    
    print(f"\n  Primes up to 50: {sieve_of_eratosthenes(50)}")
    
    # Base Conversions
    print("\n2. Base Conversions:")
    print(f"  255 in binary: {decimal_to_base(255, 2)}")
    print(f"  255 in hexadecimal: {decimal_to_base(255, 16)}")
    print(f"  255 in octal: {decimal_to_base(255, 8)}")
    print(f"  Binary 11111111 to decimal: {base_to_decimal('11111111', 2)}")
    print(f"  Hex FF to decimal: {base_to_decimal('FF', 16)}")
    
    # GCD and LCM
    print("\n3. GCD and LCM:")
    print(f"  GCD(48, 18) = {gcd(48, 18)}")
    print(f"  GCD(100, 75) = {gcd(100, 75)}")
    print(f"  LCM(12, 18) = {lcm(12, 18)}")
    
    # Maximum value for base
    print("\n4. Maximum Value for Base:")
    print(f"  Max value for base 10, 3 digits: {max_value_for_base(10, 3)}")
    print(f"  Max value for base 2, 8 digits: {max_value_for_base(2, 8)}")
    print(f"  Max value for base 16, 2 digits: {max_value_for_base(16, 2)}")
    
    # Factorial
    print("\n5. Factorial:")
    for i in range(0, 11):
        print(f"  {i}! = {factorial(i)}")
    
    # Additional algorithms
    print("\n6. Fibonacci Sequence:")
    print(f"  First 15 Fibonacci numbers: {[fibonacci(i) for i in range(15)]}")
    
    print("\n7. Fast Exponentiation:")
    print(f"  2^10 = {power(2, 10)}")
    print(f"  5^3 = {power(5, 3)}")
    print(f"  2^-3 = {power(2, -3)}")
